ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/Model.hs

```diff
@@ -1,109 +1,325 @@
 
-putBlockModel :: MonadState (DBModel blockId) m
+putBlockModel :: MonadState (MyState blockId) m
               => Ord blockId
               => ErrorHandling (VolatileDBError blockId) m
+              -> Int
+              -> (blockId -> Slot)
               -> blockId
               -> Builder
               -> m ()
-putBlockModel err sl bs = do
-    dbm@DBModel {..} <- get
+putBlockModel err maxNumPerFile toSlot bid bs = do
+    let managesToPut errors = do
+            errs <- errors
+            (mErr, _rest) <- uncons $ getStream (_hPut errs)
+            (fsErr, _mCorr) <- mErr
+            return fsErr
+    (dbm@DBModel {..}, cmdErr) <- get
     if not open then EH.throwError err ClosedDBError
-    else case Map.lookup sl mp of
-        Nothing -> put dbm{mp = Map.insert sl (toStrict $ toLazyByteString bs) mp}
+    else case Map.lookup bid mp of
         Just _bs -> return ()
+        Nothing -> case managesToPut cmdErr of
+            Just fsErrT -> EH.throwError err $ FileSystemError $
+                FsError {
+                      fsErrorType = fsErrT
+                    , fsErrorPath = [currentFile]
+                    , fsErrorString = ""
+                    , fsErrorStack = EmptyCallStack
+                    , fsLimitation = False
+                }
+            Nothing -> do
+                let mp' = Map.insert bid (toStrict $ toLazyByteString bs) mp
+                    (mbid, n, bids) = fromMaybe
+                        (error "current file does not exist in index")
+                        (Map.lookup currentFile index)
+                    n' = n + 1
+                    index' = Map.insert currentFile (updateSlotNoBlockId mbid [toSlot bid], n', bid:bids) index
+                    (currentFile', index'', nextFId') =
+                        if n' == maxNumPerFile
+                        then ( Internal.filePath nextFId
+                            , Map.insertWith
+                                (\ _ _ -> (error $ "new file " <> currentFile' <> "already in index"))
+                                currentFile' newFileInfo index'
+                            , nextFId + 1)
+                        else ( currentFile
+                            , index'
+                            , nextFId)
+                putDB dbm {mp = mp', index = index'', currentFile = currentFile', nextFId = nextFId'}
 
-garbageCollectModel :: MonadState (DBModel blockId) m
+garbageCollectModel :: forall m blockId
+                     . MonadState (MyState blockId) m
                     => ErrorHandling (VolatileDBError blockId) m
                     -> Slot
                     -> m ()
 garbageCollectModel err sl = do
-    dbm@DBModel {..} <- get
+    (DBModel {..}, cmdErr) <- get
     if not open then EH.throwError err ClosedDBError
-    else put dbm {latestGarbaged = Just $ maxMaybe latestGarbaged sl}
+    else do
+        modify $ \(dbm', cErr) -> (dbm' {latestGarbaged = Just $ maxMaybe latestGarbaged sl}, cErr)
```

Comment:
Similarly here, let's add a comment that says that if we change the exact logic of how GC is implemented (for instance, deletes files in a different order), then this tests need updating also.